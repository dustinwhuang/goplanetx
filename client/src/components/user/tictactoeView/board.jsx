import React from 'react';
import PropTypes from 'prop-types';

function Square(props) {
  return (
    <button className="square" onClick={props.onClick}>
      {props.value}
    </button>
  );
}

function HotSquare(props) {
  return <button className="square" onClick={props.onClick}>{props.value}</button>;
}

class Board extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      hotSquareClickCount: 0,
    };
  }

  resetHotSquareClickCount() {
    this.state.hotSquareClickCount = 0;
  }

  incrementHotSquareClickCount() {
    this.state.hotSquareClickCount += 1;
    if (this.state.hotSquareClickCount === 10) {
      this.props.unlockForms();
      this.state.hotSquareClickCount = 0;
    }
  }

  renderSquare(i) {
    return (
      <Square
        value={this.props.squares[i]}
        onClick={() => this.props.onClick(i)}
      />
    );
  }

  renderHotSquare(i) {
    return (
      <HotSquare
        value={this.props.squares[i]}
        onClick={() => {
          this.incrementHotSquareClickCount();
          this.props.onClick(i);
        }}
      />);
  }

  render() {
    return (
      <div>
        <div className="board-row">
          {this.renderSquare(0)}
          {this.renderSquare(1)}
          {this.renderSquare(2)}
        </div>
        <div className="board-row">
          {this.renderSquare(3)}
          {this.renderSquare(4)}
          {this.renderSquare(5)}
        </div>
        <div className="board-row">
          {this.renderHotSquare(6)}
          {this.renderSquare(7)}
          {this.renderSquare(8)}
        </div>
      </div>
    );
  }
}

Square.propTypes = {
  onClick: PropTypes.func.isRequired,
  value: PropTypes.string.isRequired,
};

HotSquare.propTypes = {
  onClick: PropTypes.func.isRequired,
  value: PropTypes.string.isRequired,
};

Board.propTypes = {
  onClick: PropTypes.func.isRequired,
  value: PropTypes.string.isRequired,
  unlockForms: PropTypes.func.isRequired,
  squares: PropTypes.shape([]).isRequired,
};

export default Board;