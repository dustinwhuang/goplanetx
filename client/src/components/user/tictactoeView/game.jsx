import React from 'react';
import PropTypes from 'prop-types';
import { Container, Row, Col } from 'reactstrap';
import { Board } from './board';
import Player from './player';
import Computer from './computer';

function calculateWinner(squares) {
  const lines = [
    [0, 1, 2],
    [3, 4, 5],
    [6, 7, 8],
    [0, 3, 6],
    [1, 4, 7],
    [2, 5, 8],
    [0, 4, 8],
    [2, 4, 6],
  ];
  for (let i = 0; i < lines.length; i += 1) {
    const [a, b, c] = lines[i];
    if (squares[a] && squares[a] === squares[b] && squares[a] === squares[c]) {
      return squares[a];
    }
  }
  return null;
}

class Game extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      history: [
        {
          squares: Array(9).fill(null),
        },
      ],
      stepNumber: 0,
      xIsNext: true,
      player1wins: 0,
      player2wins: 0,
      gameInPlay: true,
    };

    this.onReset = this.onReset.bind(this);
    this.handleClick = this.handleClick.bind(this);
  }

  onReset() {
    this.setState({
      history: [
        {
          squares: Array(9).fill(null),
        },
      ],
      stepNumber: 0,
      xIsNext: true,
      gameInPlay: true,
    });
  }

  handleClick(i) {
    const history = this.state.history.slice(0, this.state.stepNumber + 1);
    const current = history[history.length - 1];
    const squares = current.squares.slice();

    if (!this.state.gameInPlay || squares[i]) {
      return;
    }

    squares[i] = this.state.xIsNext ? '╳' : '◯';
    this.setState({
      history: history.concat([
        {
          squares,
        },
      ]),
      stepNumber: history.length,
      xIsNext: !this.state.xIsNext,
    }, () => {
      const winner = calculateWinner(squares);
      if (winner === '╳') {
        this.setState({
          player1wins: this.state.player1wins + 1,
          gameInPlay: false,
        });
      } else if (winner === '◯') {
        this.setState({
          player2wins: this.state.player2wins + 1,
          gameInPlay: false,
        });
      }
    });
  }

  jumpTo(step) {
    this.setState({
      stepNumber: step,
      xIsNext: (step % 2) === 0,
    });
  }

  render() {
    const { history } = this.state;
    const current = history[this.state.stepNumber];
    const winner = calculateWinner(current.squares);

    let status;
    if (winner) {
      status = `Winner: ${winner}`;
    } else {
      status = `Next player: ${this.state.xIsNext ? '╳' : '◯'}`;
    }

    return (
      <Container>
        <Row>
          <Col sm={3}>
            <Player player="╳" wins={this.state.player1wins} />
          </Col>
          <Col md={6}>
            <div className="game">
              <div className="game-board">
                <div className="game-info">
                  <div>{status}</div>
                </div>
                <Board
                  squares={current.squares}
                  onClick={i => this.handleClick(i)}
                  unlockForms={this.props.unlockForms}
                />
                <span><button onClick={this.onReset}>reset</button></span>
              </div>
            </div>
          </Col>
          <Col sm={3}>
            {this.props.twoPlayers ? (<Player player="◯" wins={this.state.player2wins} />)
            : (<Computer wins={this.state.player2wins} />)}
          </Col>
        </Row>
      </Container>
    );
  }
}

Game.propTypes = {
  unlockForms: PropTypes.func,
  twoPlayers: PropTypes.bool.isRequired,
};

Game.defaultProps = {
  unlockForms: () => {},
};

export default Game;
