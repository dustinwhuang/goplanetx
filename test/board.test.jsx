import React from 'react';
import { shallow } from 'enzyme';
import Game from '../client/src/components/user/tictactoeView/game';
import { Board, HotSquare } from '../client/src/components/user/tictactoeView/board';

/* TEST GAME COMPONENT */

describe('Component: Game', () => {
  it('should match its empty snapshot', () => {
    const component = shallow(<Game unlockForms={() => {}} />);
    expect(component).toMatchSnapshot();
  });
});


describe('Component: Board', () => {
  const squares = Array(9).fill(null);

  const handleClick = (i) => {
    console.log('click handled: ', i);
  };

  const unlockForms = () => {
    console.log('forms have been unlocked!');
  };

  it('should match its empty snapshot', () => {
    const component = shallow((
      <Board squares={squares} onClick={handleClick} unlockForms={unlockForms} value="x" />
    ));
    expect(component).toMatchSnapshot();
  });
});

describe('Component: Hot Button', () => {
  const value = 'X';
  const handleClick = (i) => {
    console.log('click handled: ', i);
  };
  it('should match its empty snapshot', () => {
    const component = shallow(<HotSquare value={value} onClick={handleClick} />);
    expect(component).toMatchSnapshot();
  });
});
