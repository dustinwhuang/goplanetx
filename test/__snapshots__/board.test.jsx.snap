// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Board should match its empty snapshot 1`] = `
<div>
  <div
    className="board-row"
  >
    <Square
      line={false}
      onClick={[Function]}
      value={null}
    />
    <Square
      line={false}
      onClick={[Function]}
      value={null}
    />
    <Square
      line={false}
      onClick={[Function]}
      value={null}
    />
  </div>
  <div
    className="board-row"
  >
    <Square
      line={false}
      onClick={[Function]}
      value={null}
    />
    <Square
      line={false}
      onClick={[Function]}
      value={null}
    />
    <Square
      line={false}
      onClick={[Function]}
      value={null}
    />
  </div>
  <div
    className="board-row"
  >
    <Square
      line={false}
      onClick={[Function]}
      value={null}
    />
    <Square
      line={false}
      onClick={[Function]}
      value={null}
    />
    <Square
      line={false}
      onClick={[Function]}
      value={null}
    />
  </div>
</div>
`;

exports[`Component: Game should match its empty snapshot 1`] = `
<Container
  tag="div"
>
  <Row
    tag="div"
  >
    <Col
      sm={3}
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <Player
        player="╳"
        wins={0}
      />
    </Col>
    <Col
      md={6}
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <div
        className="game"
      >
        <div
          className="game-board"
        >
          <div
            className="game-info"
          >
            <div>
              Next player: ╳
            </div>
          </div>
          <Board
            line={null}
            onClick={[Function]}
            squares={
              Array [
                null,
                null,
                null,
                null,
                null,
                null,
                null,
                null,
                null,
              ]
            }
            unlockForms={[Function]}
          />
          <span>
            <button
              onClick={[Function]}
            >
              reset
            </button>
          </span>
        </div>
      </div>
    </Col>
    <Col
      sm={3}
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <Player
        player="◯"
        wins={0}
      />
    </Col>
  </Row>
</Container>
`;

exports[`Component: Hot Button should match its empty snapshot 1`] = `
<button
  className="square-pop"
  onClick={[Function]}
>
  X
</button>
`;
